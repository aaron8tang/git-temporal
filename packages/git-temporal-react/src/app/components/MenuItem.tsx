import React from 'react';
import { style } from 'app/styles';
import { TestProps } from 'app/interfaces';

export interface MenuItemProps {
  // The children are the menu content
  children: string | JSX.Element | JSX.Element[];
  onClick: (evt) => void;
  style?: string | object;
  disabled?: boolean;
}
const initialState = {
  isHovering: false,
};
type MenuItemState = Readonly<typeof initialState>;

const containerStyle = {
  _extends: ['block', 'normalText'],
  padding: 5,
  border: '1px solid white',
};
const hoverContainerStyle = {
  _extends: [containerStyle, 'selectable'],
};
const disabledContainerStyle = {
  _extends: [containerStyle, 'disabledText'],
};

export class MenuItem extends React.Component<
  MenuItemProps & TestProps,
  MenuItemState
> {
  readonly state: MenuItemState = initialState;

  render() {
    const outerStyle = this.props.disabled
      ? disabledContainerStyle
      : this.state.isHovering
        ? hoverContainerStyle
        : containerStyle;
    return (
      <div
        style={style(outerStyle, this.props.style)}
        onClick={this.props.onClick}
        onMouseOver={this.onMouseOver}
        onMouseLeave={this.onMouseLeave}
      >
        {this.props.children}
      </div>
    );
  }

  onMouseOver = () => {
    this.setState({ isHovering: true });
  };

  onMouseLeave = () => {
    this.setState({ isHovering: false });
  };
}
