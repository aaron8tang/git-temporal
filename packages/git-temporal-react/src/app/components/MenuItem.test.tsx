import React from 'react';
import { shallow } from 'enzyme';

import { MenuItem } from './MenuItem';

describe('components/MenuItem', () => {
  describe('when rendered with required props', () => {
    let wrapper;
    const onClickFn = jest.fn();
    beforeAll(() => {
      wrapper = shallow(
        <MenuItem onClick={onClickFn}>this is the a menu item</MenuItem>
      );
    });

    test('it should match snapshot (it should have rendered a label and a text element)', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
