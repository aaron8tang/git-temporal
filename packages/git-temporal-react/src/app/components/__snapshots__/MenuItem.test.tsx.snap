// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/MenuItem when rendered with required props it should match snapshot (it should have rendered a label and a text element) 1`] = `
<div
  onClick={[MockFunction]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "border": "1px solid transparent",
      "color": "#333333",
      "display": "block",
      "fontSize": 13,
      "fontWeight": "normal",
      "marginBottom": 10,
      "padding": 5,
    }
  }
>
  this is the a menu item
</div>
`;
